### PR TITLE
Handle 32-bit overflows in disk utilization stats

### PR DESCRIPTION
**Pull request by stilor**
Linux prints [1] some of the values reported in block device's `stat` sysfs file as 32-bit unsigned integers. fio interprets them as 64-bit integers when reading that sysfs file and performs further arithmetics on them in 64-bits. If the reported value overflows during fio run, a huge bogus value is reported in the "disk utilization" block instead.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/block/genhd.c#n962

### CODE REVIEW

File: diskutil.c [hunk 1]
```diff
@@ -77,6 +77,24 @@ static int get_io_ticks(struct disk_util *du, struct disk_util_stat *dus)
 	return ret != 10;
 }
 
+static uint64_t safe_32bit_diff(uint64_t nval, uint64_t oval)
+{
+	/* Linux kernel prints some of the stat fields as 32-bit integers. It is
+	 * possible that the value overflows, but since fio uses unsigned 64-bit
+	 * arithmetics in update_io_tick_disk(), it instead results in a huge
```
Comment (szaydel) [inline diskutil.c:84]: Suggest changing from `arithmetics` to `arithmetic`, sorry not grammar police. :)

File: diskutil.c [hunk 2]
```diff
@@ -77,6 +77,24 @@ static int get_io_ticks(struct disk_util *du, struct disk_util_stat *dus)
...
+	 * bogus value being added to the respective accumulating field. Just
+	 * in case Linux starts reporting these metrics as 64-bit values in the
+	 * future, check that overflow actually happens around the 32-bit
+	 * unsigned boundary; assume overflow only happens once between
+	 * successive polls.
+	 */
+	if (oval <= nval || oval >= (1ull << 32)) {
```
Comment (ankit-sam) [inline diskutil.c:91]: No need for curly braces for this `if else` condition

**Review by ankit-sam — COMMENTED** [review]
@stilor please fix the build failure

**Comment by vincentkfu** [comment]
Please change the link in the commit message to one that doesn't require login credentials.

**Comment by stilor** [comment]
> @stilor please fix the build failure

This seems to be a GCC bug. I minimized the testcase to https://godbolt.org/z/nMEGzfW7Y and it still produces this error. Reported https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111696 to GCC maintainers with a minimal testcase derived from diskutil.c.

For now, disabled the warning using GCC diagnostic pragmas around the code triggering the spurious warning. Alternatively, if you prefer I could add a test for affected compilers in configure and if the compiler fails the test, add -Wno-stringop-overflow (to prevent such spurious warnings in all files.

File: diskutil.c [hunk 3]
```diff
@@ -96,14 +113,21 @@ static void update_io_tick_disk(struct disk_util *du)
...
+	dus->s.ticks[0] += safe_32bit_diff(__dus.s.ticks[0], ldus->s.ticks[0]);
+	dus->s.ticks[1] += safe_32bit_diff(__dus.s.ticks[1], ldus->s.ticks[1]);
+	dus->s.io_ticks += safe_32bit_diff(__dus.s.io_ticks, ldus->s.io_ticks);
+	dus->s.time_in_queue +=
+			safe_32bit_diff(__dus.s.time_in_queue, ldus->s.time_in_queue);
 
 	fio_gettime(&t, NULL);
 	dus->s.msec += mtime_since(&du->time, &t);
 	memcpy(&du->time, &t, sizeof(t));
+
```
Comment (bvanassche) [inline diskutil.c:125]: Please remove this `#pragma` and change the memcpy() calls in this function into assignments. There is no reason to use memcpy() here.

**Review by bvanassche — CHANGES_REQUESTED** [review]

File: diskutil.c (hunk 3 above)
Comment (stilor) [reply diskutil.c]: Done.

File: diskutil.c [hunk 4]
```diff
@@ -96,15 +113,16 @@ static void update_io_tick_disk(struct disk_util *du)
...
-	dus->s.time_in_queue += (__dus.s.time_in_queue - ldus->s.time_in_queue);
+	dus->s.ticks[0] += safe_32bit_diff(__dus.s.ticks[0], ldus->s.ticks[0]);
+	dus->s.ticks[1] += safe_32bit_diff(__dus.s.ticks[1], ldus->s.ticks[1]);
+	dus->s.io_ticks += safe_32bit_diff(__dus.s.io_ticks, ldus->s.io_ticks);
+	dus->s.time_in_queue +=
+			safe_32bit_diff(__dus.s.time_in_queue, ldus->s.time_in_queue);
 
 	fio_gettime(&t, NULL);
 	dus->s.msec += mtime_since(&du->time, &t);
-	memcpy(&du->time, &t, sizeof(t));
```
Comment (bvanassche) [inline diskutil.c:106]: Please consider moving this change into a separate patch.

Comment (stilor) [reply diskutil.c:106]: Done.
